File: code/code1/endpointdetect.py
'''
 * @file endpointdetect.py
 * @author Jiaxu Xiao
 * @brief Endpoint detection with double thresholds
 * @version 0.1
 * @date 2024-10-13
 * 
 * @copyright Copyright (c) 2024
'''
import numpy as np
def endpoint_detect(En, Zn):
    frame_num = len(En)
    noise_begin = len(En) - 19
    noise_end = len(En) - 1
    avg_noise_en = np.mean(En[noise_begin:noise_end])
    std_noise_en = np.std(En[noise_begin:noise_end])
    avg_noise_zn = np.mean(Zn[noise_begin:noise_end])
    std_noise_zn = np.std(Zn[noise_begin:noise_end])
    MH = np.mean(En) / 4
    ML = (avg_noise_en + MH) / 4
    Zs = avg_noise_zn + 3 * std_noise_zn
    M = 3
    i = 0
    flag = True
    while flag:
        while En[i] < MH:
            if i >= frame_num - 1:
                break
            i += 1
        flag = False
        begin_1 = i
        for j in range(i + 1, i + M + 1):
            if j >= frame_num - 1:
                break
            if En[j] < MH:
                i = j + 1
                flag = True
    i = frame_num - 1
    flag = True
    while flag:
        while En[i] < MH:
            if i <= 0:
                break
            i -= 1
        flag = False
        end_1 = i
        for j in range(i - 1, i - M - 1, -1):
            if j <= 0:
                break
            if En[j] < MH:
                i = j - 1
                flag = True
    i = begin_1 - 1
    while i > 0:
        if En[i] > ML:
            i -= 1
        else:
            break
    begin_2 = i
    i = end_1 + 1
    while i < frame_num:
        if En[i] > ML:
            i += 1
        else:
            break
    end_2 = i
    i = begin_2 - 1
    while i > 0 and Zn[i] > Zs:
        i -= 1
    begin_3 = i
    i = end_2 + 1
    while i < frame_num and Zn[i] > Zs:
        i += 1
    end_3 = i
    start = begin_3
    final = end_3
    return start, final

File: code/code_gui/endpointDetect.m
function [start, final] = endpointDetect(En, Zn)
% 基于双门限法的端点检测
frameNum = length(En);

% 测量背景噪声参数, 用于选取阈值
noise_begin = length(En) - 19;
noise_end = length(En) - 1;                     % 假定环境噪声区间
avgNoiseEn = mean(En(noise_begin:noise_end));   % 计算环境噪声的平均短时能量
stdNoiseEn = std(En(noise_begin:noise_end));    % 计算环境噪声短时能量的标准差
avgNoiseZn = mean(Zn(noise_begin:noise_end));   % 环境噪声的平均过零率
stdNoiseZn = std(Zn(noise_begin:noise_end));    % 环境噪声的过零率的标准差

% 选取阈值
MH = mean(En) / 4;                  % 较高的能量阈值
ML = (avgNoiseEn + MH) / 4;         % 较低的能量阈值
Zs = avgNoiseZn + 3 * stdNoiseZn;   % 过零率阈值

% 验证帧数
M = 3;

% 初判起止点 (MH)
% begin_1
i = 1;
flag = 1;
while flag
    while En(i) < MH
        if i >= frameNum
            break;
        end
        i = i + 1;
    end
    flag = 0;
    begin_1 = i;
    % 为了防止是突发噪声，继续向后搜索 M 帧，看是否是真正的语音
    for j = (i + 1):(i + M)
        if j >= frameNum
            break;
        end
        if En(j) < MH
            % 后一帧能量低于门限，说明前一帧大概率不是语音
            i = j + 1;
            flag = 1;
        end
    end
end
% end_1
i = frameNum;
flag = 1;
while flag
    while En(i) < MH
        if i <= 1
            break;
        end
        i = i - 1;
    end
    flag = 0;
    end_1 = i;
    % 为了防止是突发噪声，继续向后搜索 M 帧，看是否是真正的语音
    for j = (i - 1):(i - M)
        if j <= 1
            break;
        end
        if En(j) < MH
            i = j - 1;
            flag = 1;
        end
    end
end

% 判定语音段 (ML)
% begin_2
i = begin_1 - 1;
while i > 1
    if En(i) > ML
        i = i - 1;
    else
        break;
    end
end
begin_2 = i;
% end_2
i = end_1 + 1;
while i < frameNum
    if En(i) > ML
        i = i + 1;
    else
        break;
    end
end
end_2 = i;

% 语音段起止点 (Zs)
% begin_3
i = begin_2 - 1;
while i > 1 && Zn(i) > Zs
    i = i - 1;
end
begin_3 = i;
% end_3
i = end_2 + 1;
while i < frameNum && Zn(i) > Zs
    i = i + 1;
end
end_3 = i;

start = begin_3;
final = end_3;
